pkg/controller/secret/secretsync.go

```diff
@@ -0,0 +1,207 @@
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	informers v1helpers.KubeInformersForNamespaces,
+	resync time.Duration,
+	eventRecorder events.Recorder) factory.Controller {
+
+	// Read secret from operator namespace and save the translated one to the operand namespace
+	secretInformer := informers.InformersFor(util.OperatorNamespace)
+	configMapInformer := informers.InformersFor(util.ConfigMapNamespace)
+	c := &SecretSyncController{
+		operatorClient:  operatorClient,
+		kubeClient:      kubeClient,
+		secretLister:    secretInformer.Core().V1().Secrets().Lister(),
+		configMapLister: configMapInformer.Core().V1().ConfigMaps().Lister(),
+		eventRecorder:   eventRecorder.WithComponentSuffix("SecretSync"),
+	}
+	return factory.New().WithSync(c.sync).ResyncEvery(resync).WithSyncDegradedOnError(operatorClient).WithInformers(
+		operatorClient.Informer(),
+		secretInformer.Core().V1().Secrets().Informer(),
+		configMapInformer.Core().V1().Secrets().Informer(),
+	).ToController("SecretSync", eventRecorder)
+}
+
+func (c *SecretSyncController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		klog.V(2).Infof("Error while getting operator state %s", err)
+		return err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		klog.V(2).Infof("Operator management state is not managed")
+		return nil
+	}
+
+	cloudSecret, err := c.secretLister.Secrets(util.OperatorNamespace).Get(util.CloudCredentialSecretName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// TODO: report error after some while?
+			klog.V(2).Infof("Waiting for secret %s from %s", util.CloudCredentialSecretName, util.OperatorNamespace)
+			return nil
+		}
+		klog.V(2).Infof("Secret listener failed to get secret details %s", err)
+		return err
+	}
+
+	cloudConfConfigMap, err := c.configMapLister.ConfigMaps(util.ConfigMapNamespace).Get(util.ConfigMapName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// TODO: report error after some while?
+			klog.V(2).Infof("Waiting for configmap %s from %s", util.CloudCredentialSecretName, util.ConfigMapNamespace)
+			return nil
+		}
+		klog.V(2).Infof("Configmap listener failed to get cm details %s", err)
+		return err
+	}
+
+	// Get the storage-secret-store secret to be created from ibm-cloud-credential secret and clod-conf configmap
+	driverSecret, err := c.translateSecret(cloudSecret, cloudConfConfigMap)
+	if err != nil {
+		klog.V(2).Infof("Error while extracting data from secret/cm %s", err)
+		return err
+	}
+	_, _, err = resourceapply.ApplySecret(c.kubeClient.CoreV1(), c.eventRecorder, driverSecret)
+	if err != nil {
+		klog.V(2).Infof("Error while creating the secret %s", err)
+		return err
+	}
+	klog.V(2).Infof("%s secret created successfully", util.IBMCSIDriverSecretName)
+	return nil
+}
+
+func (c *SecretSyncController) translateSecret(cloudSecret *v1.Secret, cloudConf *v1.ConfigMap) (*v1.Secret, error) {
```

Comment:
Please add unit test for this function. This is core of the controller and it has simple inputs and outputs.

Reply:
added